### PR TITLE
lbank v2 (Issue) (Work in Progress): 

### PR DESCRIPTION
**Pull request by DoctorSlimm**

### CODE REVIEW

File: js/lbank2.js [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+            'symbol': market['id'],
+        };
+        const response = await this.publicGetIncrDepth (this.extend (request, params));
+        const result = this.safeValue (response, 'data', {});
+        const timestamp = result['timestamp'];
+        const orderbook = this.parseOrderBook (result, symbol, timestamp);
+        return orderbook;
+    }
+
+    parseTrade (trade, market = undefined) {
```
Comment (DoctorSlimm) [inline js/lbank2.js:323]: I think the issue is in the translation from parseTrade to parseTrades

File: js/lbank2.js [hunk 2]
```diff
@@ -396,26 +396,27 @@ module.exports = class lbank2 extends Exchange {
...
-            } else {
-                pem = this.convertSecretToPem (this.secret);
-            }
-            const sign = this.binaryToBase64 (this.rsa (message, this.encode (pem), 'RS256'));
+                'echostr': echostr,
+                'signature_method': 'HmacSHA256',
+                'timestamp': timestamp,
+            }, params)));
+            const message = this.hash (authParameters).toUpperCase ();
+            const sign = this.hmac (message, this.secret);
```
Comment (frosty00) [inline js/lbank2.js]: you need to do `this.hmac (this.encode (message), this.encode (this.secret))` for it to work in python

File: js/lbank2.js [hunk 3]
```diff
@@ -396,26 +396,27 @@ module.exports = class lbank2 extends Exchange {
...
+            query = this.extend (query, params);
             body = this.urlencode (query);
-            headers = { 'Content-Type': 'application/x-www-form-urlencoded' };
+            headers = {
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'timestamp': timestamp,
+                'signature_method': 'HmacSHA256',
+                'echostr': echostr,
+            };
+            url += '?' + 'api_key=' + this.apiKey + '&' + 'sign=' + sign;
```
Comment (frosty00) [inline js/lbank2.js]: why not add these params to the urlencode function instead of manually encoding them?

File: js/lbank2.js (hunk 2 above)
Comment (DoctorSlimm) [reply js/lbank2.js]: okay thanks will add that now

File: js/lbank2.js [hunk 4]
```diff
@@ -396,26 +396,27 @@ module.exports = class lbank2 extends Exchange {
             }
         } else {
             this.checkRequiredCredentials ();
-            const query = this.keysort (this.extend ({
+            const timestamp = Date.now ();
+            const echostr = this.uuid22 () + this.uuid16 ();
```
[the remark this answers is not in the document]
Comment (DoctorSlimm) [reply js/lbank2.js:647]: because they want a uuid which is 30-40 characters long

**Comment by zgayjjf** [comment]
It would be great if this can be merged into the mainstream.

**Comment by DoctorSlimm** [comment]
@zgayjjf its still got a little way to go, due to the Lbank API offering multiple endpoints for similar functionalities is why it's taking a longer, to implement support for all of them up to the users choice in the 'options' property.

**Comment by kroitor** [comment]
@DoctorSlimm lmk if this PR is done, I'll test on my side. Thx!

**Comment by DoctorSlimm** [comment]
@kroitor i have tested it also myself too, it is done - lmk if there's anything else to change

**Comment by aFluxx** [comment]
@DoctorSlimm 

Does the new Lbank endpoints have the ability to query closed orders (fetch_my_trades)?

**Comment by DoctorSlimm** [comment]
@aFluxx I accidentally submitted the PHP and Python files, the `lbank2.js` will be  transpiled into `lbank2.php` and `lbank2.py` so any methods which are in `lbank2.js` will also be in the php and python versions

**Comment by aFluxx** [comment]
> 

Awesome! Thanks!

**Comment by aFluxx** [comment]
@kroitor & @DoctorSlimm

I was wondering. In which release will this update be included?

**Comment by aFluxx** [comment]
@DoctorSlimm 

I copied your fork into my local project to test some things out. I'm using the php version: `lbank2.php`

When I call for example `fetch_open_orders` I get the following error:

`BadRequest The required parameters can not be empty`

This is my code that calls the underlying lbank2.php function.

```
    public function fetch(Market $market)
    {
        return new lbank([
                'apiKey' => env('MIX_LBANK_API_KEY'),
                'secret' => env('MIX_LBANK_API_SECRET'),
                'uid' => env('MIX_LBANK_API_MEMO'),
                'enableRateLimit' => true,
            ])->fetch_open_orders($market->symbol_id); // for example: BTC/USDT
    }
```

As far as I'm aware the only required parameter is the symbol ID. Because `since` & `limit` can be `null`

I must be missing something, but I can't figure out what :)

**Comment by DoctorSlimm** [comment]
@aFluxx did you use only the lbank2.js file and then run 'npm run build' and then use the generated lbank2.php file? - because the lbank2.php file in this PR is not the one you should use.

**Comment by DoctorSlimm** [comment]
@aFluxx should all work now

**Comment by aFluxx** [comment]
> @aFluxx should all work now, I had forgotten to add special encoding for the RSA-key type in sign() only for the HMAC key-type, my bad
> 
> EDIT: pls let me know if it works for you now

@DoctorSlimm 

I can probably test it out tommorrow evening or friday during the day. I'll keep you posted. Thanks for looking into this!

**Comment by zgayjjf** [comment]
@aFluxx Please come here for updates.

https://github.com/ccxt/ccxt/pull/12920

**Pull request closed without merging**